### ISSUE TITLE
Feature - Bot Activity

### ISSUE THREAD

**Issue by dandin87** [KJP12/DisFabric#6]
![image](https://user-images.githubusercontent.com/9329117/230474918-b2f9144e-c708-4d51-8822-d8eaecde41aa.png)

Irunatbullets and I manage our own forks for the All of Fabric servers, and I see you have added a lot of features. The one thing I found missing from Disfabric, that FD Link did nicely, was Bot Activity to see player counts and such. So I took that and added it to my fork. Now in our case I only care about the player count, so I just have it set to that, but FD Link had it more customizable. Would be a good addition. Here is where I add it. Not sure it was the cleanest place to have it, but it works without issues.

https://github.com/dandin87/DisFabric/blob/81e7934389f29fe72cd0bbc432e64a2f675809f7/src/main/java/br/com/brforgers/mods/disfabric/listeners/DiscordEventListener.java#L32-L37

**Comment by irunatbullets** [comment]
I got this working last night with the latest changes from the-glitch-network https://github.com/irunatbullets/DisFabric/blob/b0abea0ba1c010b07d66b23bd3cb326cff596c00/src/main/java/br/com/brforgers/mods/disfabric/listeners/DiscordEventListener.java#L59.

```
public void onMessageReceived(@NotNull MessageReceivedEvent e) {
        MinecraftServer server = getServer();
        MessageChannel channel = e.getChannel();
        if(server !=null) {
            List<ServerPlayerEntity> onlinePlayers = server.getPlayerManager().getPlayerList();
            int playerNumber = onlinePlayers.size();
            int maxPlayer = server.getMaxPlayerCount();
            DisFabric.jda.getPresence().setActivity(Activity.playing(String.valueOf(playerNumber + " / " + maxPlayer)));
        }
```

I'm possibly going to remove my repo and refork the original because I need to maintain 3 versions of the mod with our changes for 1.16.5, 1.18.2 and 1.19.2. I really liked the changes made in the-glitch-network version, but I'm a bit too dumb to try to backport everything for each of our servers.

I might also refork this one and make some PRS for this issue, the ability to use roles instead of adminid, and the ability to allow bots to run commands - but thought I'd paste the code above just in case.

**Comment by Ampflower** [comment]
> I'm possibly going to remove my repo and refork the original

Fun tip: You don't need to do a full refork at all; you can just pull from another fork and Git will manage just fine.

> the ability to use roles instead of adminid

I don't believe that's quite done, but shouldn't be much to add on (compare all the member's roles with the admin ID list).

> I might also refork this one and make some PRS for this issue

If you don't mind, I can go ahead and commit for you; tho if you prefer to sign, you can make the PR instead; although I might be able to just pull from the commit hash.


Also, same as with forking; all related forks can PR to each other fine; just...
 - `git remote add the-glitch https://github.com/the-glitch-network/DisFabric.git` (or you can use `git@github.com:` if you prefer SSH)
 - `git pull the-glitch main`, assuming you named the remote `the-glitch`.
 
IntelliJ exposes the same thing in its Git UI as well; just note that you'll have to push to your fork.

**Comment by irunatbullets** [comment]
Thank you very much for explaining the git stuff, I ended up making a bit of a mess, but I won't delete anything.

I don't mind if you commit for me.

> I don't believe that's quite done, but shouldn't be much to add on (compare all the member's roles with the admin ID list).

It seemed to work, but in interested to see.

For our use case I think it would be simpler to just check that the author has the correct role because having to reboot servers if we want to remove an admin vs. just removing their 'op' role is easier (and I'm pretty lazy)

**Comment by Ampflower** [comment]
That's fair.

I'll just pull your fork's branch in and make a couple edits where needed; will rebase into 1.19.2 since that's currently the branch I'm primarily developing on while occasionally merging into 1.19.3.

**Comment by irunatbullets** [comment]
Are you happy with the allowBots option? (Edit: though I think it might need a guild defined in the config to be better - not sure)

I ended up making the branches based on the BRForgers because I was having trouble backporting to 1.19.2, 1.18.2, and 1.16.5 (the versions of the servers we use), but I really like the changes you made to disfabric.

I'm not really a Java developer and, up until last night, I was using vscode - which was a headache. I wish I had installed intellij earlier.

**Comment by Ampflower** [comment]
Seems decent to me; I'd likely add a forced check to see if it's allowed via a set of role and bot IDs so you can enforce say only this webhook or bot is actually allowed. Tho, there's not much area for that to be a useful to thing to add on aside from making sure the bot ignores itself and its webhook.

**Comment by irunatbullets** [comment]
I don't want to create an issue for this, it's just something I noticed while debugging - I haven't been able to figure out why this is happening and perhaps I'm being dense.. 

I'm running the debug server on my machine pointed at my personal discord server, and when I put a breakpoint as shown in the image, it is picking up message from one of the AOF servers on the AOF guild.

It's weird to me because there's no mention of the server or the discord channel/guild/server anywhere in the config file.

![image](https://user-images.githubusercontent.com/344908/230803689-f6ccad80-7c16-4743-9635-4e61ac188808.png)

(It's not just this fork, it's happening on the older versions too)

**Comment by Ampflower** [comment]
The breakpoint on an if statement will break point before any of the conditions are evaluated; by nature with JDA, you'll get every message sent anywhere.

I'd recommend a more or less empty test guild with a test bot if you need to breakpoint right there, since it'll stop on *any* message; alternative breakpoint just one line after starting at `String raw = ...`

If you hit 'step', it should just skip the entire if statement if it's a bot or from the wrong channel.

Step is this arrow if you need a reference as to what it is: 
![image](https://user-images.githubusercontent.com/8528256/230805131-53028ab3-de79-4859-a871-84616635cd7f.png)

**Comment by irunatbullets** [comment]
Yeah, those are the conditions I'm testing it under, a separate guild with only me and two bots, using the internal intellij Minecraft server connected to my own disfabric bot.

In my mind, I can't really see how it's seeing the events from the other: Minecraft server, discord server, and disfabric bot.

**Comment by Ampflower** [comment]
Are you sure the test bot isn't in other servers? You might want to check mutual servers.

Also, you may want to make sure you're using the correct token for the test bot in the config, if you created a new one.

**Comment by irunatbullets** [comment]
No mutual servers and it's a fresh bot with a fresh token. Perhaps there's a bug in the version of disfabric on AOF6-NA? It still doesn't make a lot of sense though.

Does the same thing happen for you?

**Comment by Ampflower** [comment]
I see what might be happening; I think it's looping back on its own output; does this occur after sending in a command that responds, or from Minecraft chat being proxied?

If so, I think that's likely what it is.
There's not much to prevent that aside from restructuring the check to have the bot check outside of the if statement; or manually evaluating whether it should breakpoint in the breakpoints model. You should be able to edit that here:
![image](https://user-images.githubusercontent.com/8528256/230876456-a569a358-17e6-4830-8313-48328e1a2e52.png)

**Comment by irunatbullets** [comment]
I was wondering if it's something to do with disfabric not attaching a guild when listening for server chats, or possibly an issue with JDA. I started looking into it, but got a bit lost on my phone. 

Edit: it feels like it should use this in some way (/me out of my depth at this point)

https://ci.dv8tion.net/job/JDA/javadoc/net/dv8tion/jda/api/entities/Guild.html

**Comment by Ampflower** [comment]
It should be attaching a guild via the message and channel objects.

DisFabric just doesn't make use of the guild for chat messages with the exception of checking for admin roles; as you need a guild there to check for roles.
